sapi/cli/php_cli_server.c

```diff
@@ -1682,23 +1664,24 @@ static int php_cli_server_client_read_request_on_header_value(php_http_parser *p
 {
 	php_cli_server_client *client = parser->data;
 	switch (client->last_header_element) {
-	case HEADER_FIELD:
-		client->current_header_value = pestrndup(at, length, 1);
-		client->current_header_value_len = length;
-		break;
-	case HEADER_VALUE:
-		{
-			size_t new_length = client->current_header_value_len + length;
-			client->current_header_value = perealloc(client->current_header_value, new_length + 1, 1);
-			memcpy(client->current_header_value + client->current_header_value_len, at, length);
-			client->current_header_value[new_length] = '\0';
-			client->current_header_value_len = new_length;
+		case HEADER_FIELD:
+			/* Previous element was the header field, create the header value */
+			client->current_header_value = zend_string_init(at, length, /* persistent */ true);
+			break;
+		case HEADER_VALUE: {
+			/* Previous element was part of header value, append content to it */
+			size_t old_length = ZSTR_LEN(client->current_header_value);
+			// TODO Release old value?
```

Comment:
Looking at the implementation of `zend_string_extend`, it returns a `realloc`ated string if the string was non-interned and the refcount was === 1, and a new string otherwise. That's honestly not great design because there's no way to know if the returned string is copied or reallocated and if the old string needs to be released or not without basically performing the same check.

In this particular case, it looks like it is never interned and the refcount is always === 1, so the old string doesn't need to be released. Maybe it would make sense to add a `ZEND_ASSERT` to be sure.

Reply:
Will add an assertion, as I was wondering why I wasn't hitting any memory leaks

Reply:
Did you also run with `ASAN`? Memory leaks are only supported for non-persistent memory tracked by ZMM.

Reply:
> Looking at the implementation of `zend_string_extend`, it returns a `realloc`ated string if the string was non-interned and the refcount was === 1, and a new string otherwise

This follows a copy on write semantic, I don't think we need to care about the original string after a `zend_string_extend()`: The other owners will release it.

Reply:
@arnaud-lb In this case we are the owner though, right? :slightly_smiling_face: If the ref-count was >1 we'd have to release it but not if the refcount is 1 as in this case, because we don't create a copy. That's what I was referring to.

Reply:
> why the memory needs to be allocated persistently in the first place

If it's used before or after `zend_execute_scripts`, the non-persistent allocator may not be started, or maybe shutdown.

Reply:
@iluuu1994 I do think that we are the only owner here. In the ref-count > 1 case, the original string's ref-count is decremented before returning the new one, so that we don't need to release it an extra time

Reply:
> > why the memory needs to be allocated persistently in the first place
> 
> If it's used before or after `zend_execute_scripts`, the non-persistent allocator may not be started, or maybe shutdown.

That's good to know, so everything needs to be persistently allocated then in the SAPI?

Reply:
Actually I may be wrong here. The allocator is fully available as soon as sapi_cli_server_startup() and is reset in php_cli_server_request_shutdown(). So it's possible to allocate non-persistent things outside of zend_execute_script as long as their lifetime doesn't extend after php_cli_server_request_shutdown().

Reply:
Well it does seem that FreeBSD is more picky about this stuff than Linux. But lets see if me not duplicating a string breaks it again.

Reply:
Nope, I do think every string here needs to be persistently allocated.
I think I will add some ``_ex`` variant of some APIs to pass a persistent flag.

Reply:
> Nope, I do think every string here needs to be persistently allocated.

I confirm. I dug more in the cli server, and what I missed originally is that it parses multiple requests concurrently. It results that parsing and script execution (and allocator shutdown) are interleaved.